chain/message_store.go

```diff
@@ -0,0 +1,75 @@
+package chain
+
+import (
+	"context"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-hamt-ipld"
+
+	"github.com/filecoin-project/go-filecoin/types"
+)
+
+// MessageReader is an interface exposing the read methods of the
+// MessageStore.
+type MessageReader interface {
+	LoadMessages(context.Context, cid.Cid) ([]*types.SignedMessage, error)
+	LoadReceipts(context.Context, cid.Cid) ([]*types.MessageReceipt, error)
+}
+
+// MessageWriter is an interface exposing the write methods of the
+// MessageStore.
+type MessageWriter interface {
+	StoreMessages(context.Context, []*types.SignedMessage) (cid.Cid, error)
+	StoreReceipts(context.Context, []*types.MessageReceipt) (cid.Cid, error)
+}
+
+// MessageStore stores and loads collections of signed messages and receipts.
+type MessageStore struct {
+	ipldStore *hamt.CborIpldStore
+}
+
+// NewMessageStore creates and returns a new store
+func NewMessageStore(cst *hamt.CborIpldStore) *MessageStore {
+	return &MessageStore{
+		ipldStore: cst,
+	}
+}
+
+// LoadMessages loads the signed messages in the collection with cid c from ipld
+// storage.
+func (ms *MessageStore) LoadMessages(ctx context.Context, c cid.Cid) ([]*types.SignedMessage, error) {
+	// TODO #1324 message collection shouldn't be a slice
+	var out []*types.SignedMessage
+	err := ms.ipldStore.Get(ctx, c, &out)
+	if err != nil {
```

Comment:
Would it be too radical to suggest `return out, err`?

Reply:
:trollface: